master/static/srv/ship_logs.py

```diff
@@ -0,0 +1,598 @@
+        backoffs = [0, 1, 5, 10, 15]
+        self.ship(data, backoffs)
+
+
+def main(
+    master_url: str,
+    cert_name: str,
+    cert_file: str,
+    metadata: Dict[str, str],
+    token: str,
+    emit_stdout_logs: bool,
+    cmd: List[str],
+    log_wait_time: int,
+) -> int:
+    logq = queue.Queue()  # type: queue.Queue
+    doneq = queue.Queue()  # type: queue.Queue
+
+    waiter_started = False
+    stdout_started = False
+    stderr_started = False
+    shipper_started = False
+
+    # Normally we like structured concurrency; i.e. a function that owns a thread must not exit
+    # until that thead has been properly cleaned up.  However, it is important that the log shipper
+    # is not allowed to keep a task container alive too long after the child process has exited.  We
+    # want to guarantee that we exit about DET_LOG_WAIT_TIME seconds after the child process exits.
+    #
+    # However, interruping a synchronous HTTP call form urllib is nearly impossible; even if you
+    # were to select() until the underlying file descriptor had something to read before calling
+    # Request.read(), there are many buffered readers in there and most likely multiple os.read()
+    # calls would occur and you'd be blocking anyway.
+    #
+    # So as an easy workaround, we set daemon=True and just exit the process if it's not done on
+    # time.
+    shipper = Shipper(master_url, token, cert_name, cert_file, logq, doneq, daemon=True)
+    shipper_timed_out = False
+
+    # Start the process or ship a special log message to the master why we couldn't.
+    try:
+        # Don't rely on Popen's standard line buffering; we want to do our own line buffering.
+        p = subprocess.Popen(
+            cmd, stdin=subprocess.DEVNULL, stdout=subprocess.PIPE, stderr=subprocess.PIPE, bufsize=0
+        )
+    except FileNotFoundError:
+        shipper.ship_special(f"FileNotFoundError executing {cmd}", metadata, emit_stdout_logs)
+        # 127 is the standard bash exit code for file-not-found.
+        return 127
+    except PermissionError:
+        # Unable to read or to execute the command.
+        shipper.ship_special(f"PermissionError executing {cmd}", metadata, emit_stdout_logs)
+        # 126 is the standard bash exit code for permission failure.
+        return 126
+    except Exception:
+        msg = f"unexpected failure executing {cmd}:\n" + traceback.format_exc()
+        shipper.ship_special(msg, metadata, emit_stdout_logs)
+        # 80 is the exit code we use to signal "ship_logs.py failed"
+        return 80
+
+    # Just for mypy.
+    assert isinstance(p.stdout, io.RawIOBase) and isinstance(p.stderr, io.RawIOBase)
+
+    try:
+        stdout = Collector(p.stdout, "stdout", emit_stdout_logs, metadata, logq, doneq)
+        stderr = Collector(p.stderr, "stderr", emit_stdout_logs, metadata, logq, doneq)
+        waiter = threading.Thread(target=lambda: doneq.put(("waiter", p.wait(), None)))
```

Comment:
I think there's some cuteness going on here, but I haven't worked out, yet, whether it's the good kind of cuteness or the bad kind. Is there any significance to the second, "blocking" argument in the `doneq.put`, or is it there solely to make the waiter thread wait until the `cmd` subprocess is done before it publishes to `doneq`?

Reply:
> or is it there solely to make the waiter thread wait until the cmd subprocess is done before it publishes to doneq

Yeah that's the whole point.  You have to get notified as soon as the process is dead, so you can start a watchdog timer, so you don't wait forever logging to a crashed master or something.

For that to work, the wait has to happen on a separate thread.

I don't see it as cute, more like "the reason that this needs to be a thread in the first place".

Reply:
Sure, I used a NamedTuple.

Also, I decided that the inline-lambda was missing the consistent error handling that the other thread classes had, and it wasn't on-par with the "consider every possible exception case, even remote possibilities" attitude of the rest of the code.  So I expanded it into a (short) subclass of threading.Thread (thank you for challenging me on that, you were right).

Side-note, and maybe you already knew this since you didn't ask for it, but the annotation I'd like to have (`doneq: typing.Queue[DoneMsg]`) isn't possible until python 3.9.

Lastly, re `error` vs `exit_code`: the answer is that in the context we're writing, nonzero exit code is not an error.  Is a return value from a user-owned script which we must preserve.  User code failing is not considered exceptional by the platform, but the Shipper failing to ship or the Collector failing to collect or the Waiter failing to wait?  Those are all exceptional circumstances and will result in a RuntimeError.

Reply:
Thanks. I like the `DoneMsg`. I think it makes the payload significantly clearer.

Also, I buy your reasoning for why `error` is special and not just another message. Thanks.